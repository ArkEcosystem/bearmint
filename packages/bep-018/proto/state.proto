syntax = "proto3";

import "google/protobuf/struct.proto";

option optimize_for = SPEED;

message BEP21MilestoneParameters {
  .google.protobuf.Struct application = 1;
  .google.protobuf.Struct consensus = 2;
  .google.protobuf.Struct denominations = 3;
  .google.protobuf.Struct mandatory_modules = 4;
  .google.protobuf.Struct transactions = 5;
}

message BEP21Milestone {
  uint64 block_number = 1;
  BEP21MilestoneParameters parameters = 2;
}
