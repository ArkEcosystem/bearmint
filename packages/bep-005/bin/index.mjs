#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
