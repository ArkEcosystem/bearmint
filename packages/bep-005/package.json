{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
