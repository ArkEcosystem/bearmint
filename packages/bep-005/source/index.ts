export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
