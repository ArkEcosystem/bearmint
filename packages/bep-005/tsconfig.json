{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
