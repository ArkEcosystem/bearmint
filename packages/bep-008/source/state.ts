import type { TxTestContext } from '@bearmint/bep-006'
import { milestone } from '@bearmint/bep-006'
import { canonicalizeJson, Paths, snakeCase } from '@bearmint/bep-009'
import type { KVStore } from '@bearmint/bep-013'
import { ContainerType } from '@bearmint/bep-013'
import { BEP21Milestone } from '@bearmint/bep-018'
import { makeMultiStore, makeState, TK_MILESTONE } from '@bearmint/bep-020'
import { makeSecureKVStore } from '@bearmint/bep-023'

import { fakeAccountRepositoryFactory } from './account.js'

interface StateStoreKVStores {
	account: KVStore
	accountMoniker: KVStore
	accountPublicKey: KVStore
	accountValidatorAddress: KVStore
	milestone: KVStore
	tx: KVStore
	txReceipt: KVStore
	world: KVStore
}

export async function fakeStateStore(context: TxTestContext) {
	context.container.bindValue(ContainerType.CommittedState, {
		getCommittedBlockAppHash() {
			return Buffer.from('deadbeef')
		},
		getMilestone() {
			return { ...milestone }
		},
	})

	context.container.bindValue(ContainerType.DeliverTxState, {
		getCommittedBlockAppHashpHash() {
			return Buffer.from('deadbeef')
		},
		getMilestone() {
			return { ...milestone }
		},
	})

	const result = await makeState(
		fakeAccountRepositoryFactory(context),
		await fakeMultiStore({
			milestone: await fakeMilestoneKVStore(),
		}),
	)

	await result.setMilestone(milestone as any)

	return result
}

export async function fakeMilestoneKVStore() {
	const result = await makeSecureKVStore().make()
	await result.set(
		TK_MILESTONE,
		BEP21Milestone.fromJson(
			canonicalizeJson({
				blockNumber: '1',
				parameters: {
					application: milestone.parameters.application,
					consensus: milestone.parameters.consensus,
					denominations: milestone.parameters.denominations,
					mandatoryModules: milestone.parameters.modules.mandatory,
					transactions: milestone.parameters.transactions,
				},
			}),
		).toBinary(),
	)

	return result
}

export async function fakeKVStores(tries?: Partial<StateStoreKVStores>) {
	return {
		account: tries?.account ?? (await makeSecureKVStore().make()),
		accountMoniker: tries?.accountMoniker ?? (await makeSecureKVStore().make()),
		accountPublicKey: tries?.accountPublicKey ?? (await makeSecureKVStore().make()),
		accountValidatorAddress: tries?.accountValidatorAddress ?? (await makeSecureKVStore().make()),
		milestone: tries?.milestone ?? (await makeSecureKVStore().make()),
		tx: tries?.tx ?? (await makeSecureKVStore().make()),
		txReceipt: tries?.txReceipt ?? (await makeSecureKVStore().make()),
		world: tries?.world ?? (await makeSecureKVStore().make()),
	}
}

export async function fakeMultiStore(tries?: Partial<StateStoreKVStores>) {
	const result = makeMultiStore({
		KVStoreFactory: makeSecureKVStore(),
		// @ts-ignore
		Logger: {
			debug() {
				//
			},
		},
		// @ts-ignore
		Paths: {
			data: `${Paths.BM_PATH_HOME}/data`,
		},
	})

	for (const [key, trie] of Object.entries(await fakeKVStores(tries))) {
		result.set(`@bearmint/bep-020/${snakeCase(key)}`, trie)
	}

	return result
}
