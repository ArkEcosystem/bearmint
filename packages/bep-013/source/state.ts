import type { abci, BEP88Validator } from '@bearmint/bep-018'

import type { Account, AccountRepository, AccountWithValidator } from './account.js'
import type { Milestone } from './milestone.js'
import type { KVStore, KVStoreOpts } from './store.js'

export interface StateStore {
	// Mutation
	copy(): Promise<StateStore>

	// KVStore Checkpoints
	hasCheckpoints(): boolean
	checkpoint(): Promise<void>
	commit(candidateState?: StateStore | undefined): Promise<{ height: bigint; root: Uint8Array }>
	revert(): Promise<void>

	// Stores
	getMultiStore(): MultiStore
	getTxReceiptStore(): KVStore
	getTxStore(): KVStore

	// Accounts
	getAccountRepository(): AccountRepository
	setAccountRepository(values: AccountRepository): void

	// Gas Consumption
	setConsumedGas(gas: bigint): void
	getConsumedGas(): bigint

	// Milestones
	getMilestone<T = object>(): Milestone<T>
	setMilestone(milestone: Milestone): Promise<void>

	// The root of the
	getAppHash(): Uint8Array

	// A candidate block is a block that has not yet reached consensus
	hasCandidateBlock(): boolean
	getCandidateBlock(): abci.RequestFinalizeBlock
	setCandidateBlock(proposedBlock: abci.RequestFinalizeBlock): void
	getCandidateBlockHash(): Uint8Array
	getCandidateBlockNumber(): bigint
	getCandidateBlockProposer(): Uint8Array

	// A committed block is a block that has reached consensus and has been persisted
	getCommittedBlockNumber(): Promise<bigint>
	getCommittedBlock(): Promise<abci.RequestFinalizeBlock>
	setCommittedBlock(block: abci.RequestFinalizeBlock): Promise<void>
	getCommittedBlockAppHash(): Promise<Uint8Array>

	// The last set of validator updates
	getValidatorUpdates(): Promise<abci.ValidatorUpdate[]>
	setValidatorUpdates(updates: abci.ValidatorUpdate[]): Promise<void>
}

export interface AccountRepositoryFactory {
	make(tries: {
		account: KVStore
		accountMoniker: KVStore
		accountPublicKey: KVStore
		accountValidatorAddress: KVStore
	}): AccountRepository
}

export interface MultiStore {
	add(module: string, opts?: KVStoreOpts): Promise<void>
	all(): Record<string, KVStore>
	checkpoint(): void
	copy(): MultiStore
	get(module: string): KVStore
	hasCheckpoints(): boolean
	revert(): Promise<void>
	roots(): Record<string, Buffer>
	set(module: string, trie: KVStore): void
}

export interface MultiStoreFactory {
	make(): MultiStore
}

export interface StakingService {
	allByValidator(validator: AccountWithValidator): Promise<BEP88Validator>
	keysByValidator(validator: AccountWithValidator): Promise<string[]>
	valuesByValidator(validator: AccountWithValidator): Promise<bigint[]>

	increaseStake(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	decreaseStake(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	multiplyStake(
		validator: AccountWithValidator,
		delegator: Account,
		percentage: number,
	): Promise<void>

	attach(validator: AccountWithValidator, delegator: Account, stake: bigint): Promise<void>
	detach(validator: AccountWithValidator, delegator: Account): Promise<void>
}

export interface StakingServiceFactory {
	make(state: StateStore): StakingService
}
