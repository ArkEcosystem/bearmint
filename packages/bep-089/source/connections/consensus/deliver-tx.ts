import { assert } from '@bearmint/bep-009'
import type { ABCIController, Cradle, DataSink, Tx, TxReceipt } from '@bearmint/bep-013'
import { Event, TxReceiptModelSubType, TxReceiptModelType } from '@bearmint/bep-013'
import type { abci } from '@bearmint/bep-018'
import { TxFailuresRecordedException } from '@bearmint/bep-109'

function receiptToDataSink(tx: Tx, receipt: TxReceipt, dataSink: DataSink) {
	for (const log of receipt.logs()) {
		const json = log.args?.toJson() as Record<string, string>

		if (json['_module'] === undefined) {
			continue
		}

		if (json['_modelKey'] === undefined) {
			continue
		}

		if (json['_modelType'] === undefined) {
			continue
		}

		if (json['_modelSubType'] === undefined) {
			continue
		}

		dataSink.put(
			json['_module'],
			{
				id: json['_modelKey'],
				subType: json['_modelSubType'],
				type: json['_modelType'],
			},
			{ key: json['key'], value: json['value'] },
		)
	}

	// We also want to index the receipt itself
	dataSink.put(
		tx.checksum.handler,
		{
			id: tx.hash,
			subType: TxReceiptModelSubType.Receipt,
			type: TxReceiptModelType.Tx,
		},
		receipt.toJson(),
	)
}

/**
 * @deprecated going to be replaced by "finalizeBlock"
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/apps.md#consensus-connection
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/apps.md#delivertx
 * @see https://github.com/tendermint/tendermint/blob/v0.34.x/spec/abci/abci.md#delivertx-2
 */
export function makeDeliverTx(
	cradle: Cradle,
): ABCIController<abci.RequestDeliverTx, abci.ResponseDeliverTx> {
	return {
		async execute(request) {
			let tx: Tx

			try {
				assert.defined<Uint8Array>(request.tx)

				tx = await cradle.TxFactory.fromBytes(request.tx)
			} catch (error) {
				return {
					code: 1,
					data: request.tx,
					events: [],
					gasUsed: BigInt('0'),
					gasWanted: BigInt('0'),
					info: error.message,
				}
			}

			try {
				await cradle.DeliverTxState.checkpoint()

				const gasMeter = cradle.GasMeterFactory.make()

				const { events, receipt } = await cradle.TxProcessor.execute(
					cradle.EventRecorderFactory.make(),
					gasMeter,
					cradle.DeliverTxState,
					tx,
				)

				cradle.DeliverTxState.setConsumedGas(
					cradle.DeliverTxState.getConsumedGas() + gasMeter.total(),
				)

				await cradle.DeliverTxState.getTxStore().set(tx.hash, tx.bytes)
				await cradle.DeliverTxState.getTxReceiptStore().set(tx.hash, receipt.toBinary())

				// Send all receipt logs to the data sink
				receiptToDataSink(tx, receipt, cradle.DataSink)

				await cradle.EventDispatcher.dispatch(Event.TransactionIndexed, tx)

				/* eslint-disable sort-keys-fix/sort-keys-fix */
				cradle.Logger.debug('%s', {
					module: '@bearmint/bep-089',
					action: 'commit_deliver_tx_state',
					hash: tx.hash,
				})
				/* eslint-enable sort-keys-fix/sort-keys-fix */

				await cradle.DeliverTxState.commit()

				return {
					code: 0,
					data: tx.bytes,
					events,
					gasUsed: tx.data.gas,
					gasWanted: (await cradle.GasCalculator.execute(cradle.DeliverTxState, tx)).expected,
				}
			} catch (error) {
				await cradle.DeliverTxState.revert()

				/* eslint-disable sort-keys-fix/sort-keys-fix */
				cradle.Logger.debug('%s', {
					module: '@bearmint/bep-089',
					action: 'revert_deliver_tx_state',
					hash: tx.hash,
				})
				/* eslint-enable sort-keys-fix/sort-keys-fix */

				const events: abci.Event[] = []

				if (error instanceof TxFailuresRecordedException) {
					events.push(error.event)
				}

				cradle.Logger.error(error.stack)

				return {
					code: 1,
					data: tx.bytes,
					events,
					gasUsed: BigInt('0'),
					gasWanted: (await cradle.GasCalculator.execute(cradle.DeliverTxState, tx)).expected,
					info: error.message,
				}
			}
		},
	}
}
