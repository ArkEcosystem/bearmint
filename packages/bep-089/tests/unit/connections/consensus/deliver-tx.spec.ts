import { expect, describe, it, beforeEach } from 'vitest'
import { extendExpect, spy, spyFn } from '@bearmint/bep-005'
extendExpect(expect)

import { makeContainer } from '@bearmint/bep-014'
import { makeEventDispatcher } from '@bearmint/bep-015'
import { makeEventRecorderFactory } from '@bearmint/bep-090'
import { TxFailuresRecordedException } from '@bearmint/bep-109'
import { makeGasMeterFactory } from '@bearmint/bep-048'

import { fakeStateStore } from '../../../state.js'
import { makeDeliverTx } from '../../../../source/connections/consensus/deliver-tx.js'
import { abci } from '@bearmint/bep-018'

describe('DeliverTx', () => {
	it('it should respond without any errors', async () => {
		const DeliverTxState = await fakeStateStore()
		DeliverTxState.setCandidateBlock(new abci.RequestBeginBlock({ header: { height: BigInt(1) } }))
		await DeliverTxState.setCommittedBlock(
			new abci.RequestBeginBlock({ header: { height: BigInt(1) } }),
		)

		const TxProcessor = spyFn()

		expect(
			await makeDeliverTx({
				Container: makeContainer(),
				DataSink: { put() {} },
				DeliverTxState,
				EventDispatcher: makeEventDispatcher(),
				EventRecorderFactory: makeEventRecorderFactory(),
				GasCalculator: {
					execute() {
						return {
							expected: BigInt(5),
						}
					},
				},
				GasMeterFactory: makeGasMeterFactory(),
				Logger: console,
				TxFactory: {
					fromBytes() {
						return {
							bytes: Buffer.alloc(0),
							checksum: {},
							data: {
								gas: BigInt(5),
								hash: Buffer.alloc(32).toString('hex'),
							},
						}
					},
				},
				TxProcessor: {
					execute() {
						TxProcessor.call()

						return {
							events: [],
							receipt: {
								logs() {
									return []
								},
								toJson() {
									return {}
								},
								toBinary() {
									return Buffer.alloc(0)
								},
							},
						}
					},
				},
			}).execute({
				tx: Buffer.from(''),
			}),
		).toMatchSnapshot()

		TxProcessor.calledOnce()
	})

	it('it should respond after handling an unexpected error', async () => {
		const DeliverTxState = await fakeStateStore()

		const TxProcessor = spyFn()
		const TxAuditorRegistry = spyFn()
		const revert = spy(DeliverTxState, 'revert')

		expect(
			await makeDeliverTx({
				Container: makeContainer(),
				DeliverTxState,
				EventDispatcher: makeEventDispatcher(),
				EventRecorderFactory: makeEventRecorderFactory(),
				GasCalculator: {
					execute() {
						return {
							expected: BigInt(5),
						}
					},
				},
				GasMeterFactory: makeGasMeterFactory(),
				Logger: console,
				TxFactory: {
					fromBytes() {
						return {
							checksum: {},
							data: {
								gas: BigInt(5),
							},
						}
					},
				},
				TxProcessor: {
					execute() {
						throw new Error('hello')
					},
				},
			}).execute({
				tx: Buffer.from(''),
			}),
		).toStrictEqual({
			code: 1,
			data: undefined,
			events: [],
			gasUsed: BigInt('0'),
			gasWanted: BigInt('5'),
			info: 'hello',
		})

		TxAuditorRegistry.neverCalled()
		TxProcessor.neverCalled()
		revert.calledOnce()
	})

	it('it should respond after handling TxFailuresRecordedException', async () => {
		const DeliverTxState = await fakeStateStore()

		const TxProcessor = spyFn()
		const TxAuditorRegistry = spyFn()
		const revert = spy(DeliverTxState, 'revert')

		expect(
			await makeDeliverTx({
				Container: makeContainer(),
				DeliverTxState,
				EventDispatcher: makeEventDispatcher(),
				EventRecorderFactory: makeEventRecorderFactory(),
				GasCalculator: {
					execute() {
						return {
							expected: BigInt(5),
						}
					},
				},
				GasMeterFactory: makeGasMeterFactory(),
				Logger: console,
				TxFactory: {
					fromBytes() {
						return {
							checksum: {},
							data: {
								gas: BigInt(5),
							},
						}
					},
				},
				TxProcessor: {
					execute() {
						throw new TxFailuresRecordedException('hash', {
							attributes: ['attribute'],
						})
					},
				},
			}).execute({
				tx: Buffer.from(''),
			}),
		).toStrictEqual({
			code: 1,
			data: undefined,
			events: [{ attributes: ['attribute'] }],
			gasUsed: BigInt('0'),
			gasWanted: BigInt('5'),
			info: 'Tx (hash) contained (1) errors',
		})

		TxAuditorRegistry.neverCalled()
		TxProcessor.neverCalled()
		revert.calledOnce()
	})
})
