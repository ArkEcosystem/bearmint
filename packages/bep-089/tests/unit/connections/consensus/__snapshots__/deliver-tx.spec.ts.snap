// Vitest Snapshot v1

exports[`DeliverTx > it should respond without any errors 1`] = `
{
  "code": 0,
  "data": {
    "data": [],
    "type": "Buffer",
  },
  "events": [],
  "gasUsed": 5n,
  "gasWanted": 5n,
}
`;
